resources_portal/models/organization_user_permission.py

```diff
@@ -0,0 +1,22 @@
+from django.contrib.auth.models import Permission
+from django.db import models
+
+from .organization import Organization
+from .user import User
+
+
+class OrganizationUserPermission(models.Model):
+    class Meta:
+        db_table = "organization_user_permission"
+        get_latest_by = "created_at"
+
+    objects = models.Manager()
+
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    user = models.ForeignKey(User, blank=False, null=False, on_delete=models.CASCADE)
+    organization = models.ForeignKey(
+        Organization, blank=False, null=False, on_delete=models.CASCADE, related_name="permissions"
+    )
+    permission = models.ForeignKey(Permission, blank=False, null=False, on_delete=models.CASCADE)
```

Comment:
I'm not sure the built-in Django permission is going to be able to do what we need. We want to be able to assign users permissions to do specific actions, and these will be organization specific. After a quick glance at the Django permissions, it looks like they allow add, modify, delete permissions for types of objects or specific instances of objects. I don't know that we'll be able to easily encode abstract concepts like "allowed to approve requests for this organization" into that system.

Reply:
I think those are the default permissions. I have never tried it but maybe something like [this](https://docs.djangoproject.com/en/3.0/topics/auth/default/#programmatically-creating-permissions) could work:

```py
content_type = ContentType.objects.get_for_model(Organization)
permission = Permission.objects.create(
    codename='approve_requests',
    name='Can approve requests',
    content_type=content_type,
)
```

And then check if a user has that permission with [has_perm](https://docs.djangoproject.com/en/3.0/ref/contrib/auth/#django.contrib.auth.models.User.has_perm):

```
user.has_perm('resources_portal.approve_requests')
```

Do you think something like that could work?

Reply:
I guess my issue is, how does `user.has_perm('resources_portal.approve_requests')` know for which orgs the user has that perm?

Reply:
good point, hmm... that's logic that we will have to include in any case right? when checking if a user has permissions to approve requests for a specific organization? I was thinking something like:

```
user.has_perm('resources_portal.approve_requests') and \
OrganizationUserPermission.objects.filter(
  user=user, 
  permission__codename="approve_requests", 
  organization=some_org
).exists()
```

It looks like there's some redundant info there. Even if we had our own `Permission` models we would have to do something similar right?

Reply:
```
OrganizationUserPermission.objects.filter(
  user=user, 
  permission__codename="approve_requests", 
  organization=some_org
).exists()
```
Seems like what I thinking we'd need. What does `user.has_perm('resources_portal.approve_requests')` add on top of that?

Reply:
yeah, it doesn't add much. Why would we want to roll our version of `Permission`? The model we had is very similar to the one provided by Django which as a bonus integrates with the admin interface.

Also, I just found `django-guardian`, which addresses [this exact problem](https://django-guardian.readthedocs.io/en/stable/userguide/assign.html#for-user). Would this work?

Reply:
We talked and decided to try `django-guardian`, which means we'll no longer need `OrganizationUserPermission`. I'll also add some tests to show how the permissions work both with a specific `Permission` and `Group`.

Reply:
I think it'll work very well, added a test with the following:

```
        user = UserFactory.create()
        organization = OrganizationFactory.create()

        self.assertFalse(user.has_perm("approve_requests", organization))

        self.assertTrue(organization.owner.has_perm("approve_requests", organization))
```

Used a `post_save` signal on `Organization` to assign the default permissions to the organization's owner. We can do something similar with the other permissions that we use.